### DISCUSSION TITLE
Update source code

### DISCUSSION THREAD

**Discussion by FedericoTartarini** [CenterForTheBuiltEnvironment/pythermalcomfort#109] [General]
This discussion aims to establish a procedure for handling code changes within the pythermalcomfort package. I will explore various scenarios that we may encounter and I would like your opinion on how to ensure a transparent and effective development process.

Scenarios:

1. Scientific Correction: A published paper identifies an error in a model coefficient.
2. Equation Error: A paper reveals inaccuracies in specific equations within a model.
3. Significant Code Change: A paper proposes substantial modifications to the underlying codebase.
4. Standard Updates: A new standard introduces changes relevant to the model code.

Discussion Points:

1. Implementation: Should we incorporate these changes into pythermalcomfort?

* Consider the impact on existing users and potential disruptions.
* How to evaluate the credibility and significance of the proposed changes.

2. Versioning:

* When to update the underlying code?
* For critical errors (e.g., scenario 1), a prompt update might be necessary.
* For substantial changes (scenarios 2 & 3), consider a major version bump with clear migration guides.
* When to release a new version?

3. Voting System:

* Is a voting system suitable for deciding on code changes, similar to ASHRAE 55?
* How to define voting rights based on factors like:
  - Developer contributions
  - Expertise in relevant thermal comfort models
  - Standard Interaction:

4. How to handle changes in a model included within a standard?
* Advocate for updates to the standard if feasible.
* If standard updates are not immediate, consider implementing the changes in pythermalcomfort with clear communication to users and standard developers.

@stefanoschiavon @praftery @charliehuizenga @tom-parkinson @t-kramer I would like to know your opinion. Please also feel free to share a link to this discussion with other people.

**Comment by tom-parkinson** [comment]
Thanks for pulling this together @FedericoTartarini. My thoughts below: 

- I've seen this successfully managed in other open source project in a number of ways. The one I like the most is with protected branches that [use pull requests with approvals](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews). That also partially address your 'voting' question and shifts some of the burden across team members.
- I don't see the harm in responding to updates in the way that you've mentioned: minor version bumps for bug fixes and major version bumps for substantial changes or new features. Commits clearly documented in [releases](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) will allow the user to know what changes have been made and which version they'd like to use. 
- I don't think you need to define the frequency of updates; they are published when needed / possible.
- No standard updates are immediate 😄 I like the way you support multiple standards in pythermalcomfort using arguments in functions. I recognise it's more work, but it is one way of handling the transition from new to old.

**Comment by t-kramer** [comment]
@FedericoTartarini, my thoughts on your discussion points below:
1. There is probably no straight answer to this and it very much depends on the case. For the mentioned reasons (impact on users, credibility etc.), I think we should be cautious with adjusting model coefficients straight after a new publication, unless the improvement is obvious. Generally, I would trust the scientific process here, solid research findings will be confirmed by other works at some point, which then provides enough reasoning to update the model coefficients.
2. I agree with @tom-parkinson's points here. Minor version bumps for small changes (e.g. changing coefficients) and major version bumps only for substantial changes like new features.
3. I don't think we need a voting system. I think we can the trust our own judgement and the vast experience of folks in our corner (@charliehuizenga @stefanoschiavon @praftery etc.) to discuss certain points if necessary. Generally, I think expertise in thermal comfort should have more weight than developer contributions when making these decisions.
4. Again, I am with Tom. But I think we need to clearly communicate which sets of arguments follow current standards like ASHRAE 55 and which ones are based on recent scientific publications not yet implemented into standards. I'd argue that most people use the inputs that follow a standard like ASHRAE 55, and not recent scientific findings.
